File: src/main/java/agentdecompile/util/SchemaUtil.java
/* ###
 * IP: AgentDecompile
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package agentdecompile.util;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.modelcontextprotocol.spec.McpSchema;

/**
 * Utility methods for creating MCP JSON schemas.
 * <p>
 * MCP SDK references:
 * <ul>
 *   <li>{@link io.modelcontextprotocol.spec.McpSchema.JsonSchema} - MCP schema for tool parameters</li>
 *   <li>MCP Java SDK: <a href="https://github.com/modelcontextprotocol/java-sdk">MCP Java SDK</a></li>
 *   <li>MCP Server docs: <a href="https://modelcontextprotocol.info/docs/sdk/java/mcp-server/">MCP Java Server</a></li>
 * </ul>
 * </p>
 */
public class SchemaUtil {
    /**
     * Create a string property schema
     * @param description Description of the property
     * @return Map representation of the property schema
     */
    public static Map<String, Object> stringProperty(String description) {
        return Map.of(
            "type", "string",
            "description", description
        );
    }

    /**
     * Create a string property schema (alias for consistency)
     * @param description Description of the property
     * @return Map representation of the property schema
     */
    public static Map<String, Object> createStringProperty(String description) {
        return stringProperty(description);
    }

    /**
     * Create a string property schema with a default value
     * @param description Description of the property
     * @param defaultValue Default value for the property
     * @return Map representation of the property schema
     */
    public static Map<String, Object> stringPropertyWithDefault(String description, String defaultValue) {
        return Map.of(
            "type", "string",
            "description", description,
            "default", defaultValue
        );
    }

    /**
     * Create an optional string property schema (alias)
     * @param description Description of the property
     * @return Map representation of the property schema
     */
    public static Map<String, Object> createOptionalStringProperty(String description) {
        return stringProperty(description);
    }

    /**
     * Create a boolean property schema
     * @param description Description of the property
     * @return Map representation of the property schema
     */
    public static Map<String, Object> booleanProperty(String description) {
        return Map.of(
            "type", "boolean",
            "description", description
        );
    }

    /**
     * Create an optional boolean property schema (alias)
     * @param description Description of the property
     * @return Map representation of the property schema
     */
    public static Map<String, Object> createOptionalBooleanProperty(String description) {
        return booleanProperty(description);
    }

    /**
     * Create a boolean property schema with a default value
     * @param description Description of the property
     * @param defaultValue Default value for the property
     * @return Map representation of the property schema
     */
    public static Map<String, Object> booleanPropertyWithDefault(String description, boolean defaultValue) {
        return Map.of(
            "type", "boolean",
            "description", description,
            "default", defaultValue
        );
    }

    /**
     * Create an integer property schema
     * @param description Description of the property
     * @return Map representation of the property schema
     */
    public static Map<String, Object> integerProperty(String description) {
        return Map.of(
            "type", "integer",
            "description", description
        );
    }

    /**
     * Create a number property schema
     * @param description Description of the property
     * @return Map representation of the property schema
     */
    public static Map<String, Object> createNumberProperty(String description) {
        return integerProperty(description);
    }

    /**
     * Create an optional number property schema
     * @param description Description of the property
     * @return Map representation of the property schema
     */
    public static Map<String, Object> createOptionalNumberProperty(String description) {
        return integerProperty(description);
    }

    /**
     * Create an integer property schema with a default value
     * @param description Description of the property
     * @param defaultValue Default value for the property
     * @return Map representation of the property schema
     */
    public static Map<String, Object> integerPropertyWithDefault(String description, int defaultValue) {
        return Map.of(
            "type", "integer",
            "description", description,
            "default", defaultValue
        );
    }

    /**
     * Create an object property schema
     * @param description Description of the property
     * @param properties Properties of the object
     * @return Map representation of the property schema
     */
    public static Map<String, Object> createOptionalObjectProperty(String description, Map<String, Object> properties) {
        Map<String, Object> schema = new HashMap<>();
        schema.put("type", "object");
        schema.put("description", description);
        schema.put("properties", properties);
        return schema;
    }

    /**
     * Create a JSON schema object
     * @param properties Map of property names to property schemas
     * @param required List of required property names
     * @return JsonSchema object
     */
    public static McpSchema.JsonSchema createSchema(Map<String, Object> properties, List<String> required) {
        // Allow additional properties to be ignored rather than causing validation errors
        return new McpSchema.JsonSchema("object", properties, required, true, null, null);
    }

    /**
     * Create a schema builder to fluently build a schema
     * @return A new schema builder
     */
    public static SchemaBuilder builder() {
        return new SchemaBuilder();
    }

    /**
     * Builder class for creating schemas
     */
    public static class SchemaBuilder {
        private final Map<String, Object> properties = new HashMap<>();
        private final List<String> required = new java.util.ArrayList<>();

        private SchemaBuilder() {
            // Private constructor to force use of SchemaUtil.builder()
        }

        /**
         * Add a string property
         * @param name Property name
         * @param description Property description
         * @return This builder for method chaining
         */
        public SchemaBuilder stringProperty(String name, String description) {
            properties.put(name, SchemaUtil.stringProperty(description));
            return this;
        }

        /**
         * Add a string property with a default value
         * @param name Property name
         * @param description Property description
         * @param defaultValue Default value
         * @return This builder for method chaining
         */
        public SchemaBuilder stringProperty(String name, String description, String defaultValue) {
            properties.put(name, SchemaUtil.stringPropertyWithDefault(description, defaultValue));
            return this;
        }

        /**
         * Add a boolean property
         * @param name Property name
         * @param description Property description
         * @return This builder for method chaining
         */
        public SchemaBuilder booleanProperty(String name, String description) {
            properties.put(name, SchemaUtil.booleanProperty(description));
            return this;
        }

        /**
         * Add a boolean property with a default value
         * @param name Property name
         * @param description Property description
         * @param defaultValue Default value
         * @return This builder for method chaining
         */
        public SchemaBuilder booleanProperty(String name, String description, boolean defaultValue) {
            properties.put(name, SchemaUtil.booleanPropertyWithDefault(description, defaultValue));
            return this;
        }

        /**
         * Add an integer property
         * @param name Property name
         * @param description Property description
         * @return This builder for method chaining
         */
        public SchemaBuilder integerProperty(String name, String description) {
            properties.put(name, SchemaUtil.integerProperty(description));
            return this;
        }

        /**
         * Add an integer property with a default value
         * @param name Property name
         * @param description Property description
         * @param defaultValue Default value
         * @return This builder for method chaining
         */
        public SchemaBuilder integerProperty(String name, String description, int defaultValue) {
            properties.put(name, SchemaUtil.integerPropertyWithDefault(description, defaultValue));
            return this;
        }

        /**
         * Add a required property
         * @param name Property name
         * @return This builder for method chaining
         */
        public SchemaBuilder required(String name) {
            required.add(name);
            return this;
        }

        /**
         * Build the schema
         * @return JsonSchema object
         */
        public McpSchema.JsonSchema build() {
            return SchemaUtil.createSchema(properties, required);
        }
    }
}

File: src/agentdecompile_cli/mcp_utils/schema_util.py
"""Schema utility helpers for MCP JSON schema creation.

SchemaUtil builds MCP tool inputSchema property dicts (string, boolean, integer,
enum, etc.) with optional default/enum/items. Used by providers that construct
schemas programmatically instead of inline dicts. _UNSET is the sentinel for
"no default" so we can distinguish default=None from omitted.
"""

from __future__ import annotations

import logging
logger = logging.getLogger(__name__)

from typing import Any

_UNSET = object()


class SchemaUtil:
    """Utility methods for creating MCP JSON schemas."""

    @staticmethod
    def _property_schema(
        schema_type: str,
        description: str,
        *,
        default: Any = _UNSET,
        enum: list[str] | None = None,
        items: dict[str, Any] | None = None,
        properties: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Create a JSON-schema property with optional common fields."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil._property_schema")
        schema: dict[str, Any] = {
            "type": schema_type,
            "description": description,
        }
        if default is not _UNSET:
            schema["default"] = default
        if enum:
            schema["enum"] = enum
        if items:
            schema["items"] = items
        if properties:
            schema["properties"] = properties
        return schema

    @staticmethod
    def string_property(description: str) -> dict[str, Any]:
        """Create a string property schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.string_property")
        return SchemaUtil._property_schema("string", description)

    @staticmethod
    def string_property_with_default(description: str, default_value: str) -> dict[str, Any]:
        """Create a string property schema with a default value."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.string_property_with_default")
        return SchemaUtil._property_schema("string", description, default=default_value)

    @staticmethod
    def optional_string_property(description: str) -> dict[str, Any]:
        """Create an optional string property schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.optional_string_property")
        return SchemaUtil.string_property(description)

    @staticmethod
    def boolean_property(description: str) -> dict[str, Any]:
        """Create a boolean property schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.boolean_property")
        return SchemaUtil._property_schema("boolean", description)

    @staticmethod
    def boolean_property_with_default(description: str, default_value: bool) -> dict[str, Any]:
        """Create a boolean property schema with a default value."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.boolean_property_with_default")
        return SchemaUtil._property_schema("boolean", description, default=default_value)

    @staticmethod
    def optional_boolean_property(description: str) -> dict[str, Any]:
        """Create an optional boolean property schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.optional_boolean_property")
        return SchemaUtil.boolean_property(description)

    @staticmethod
    def integer_property(description: str) -> dict[str, Any]:
        """Create an integer property schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.integer_property")
        return SchemaUtil._property_schema("integer", description)

    @staticmethod
    def integer_property_with_default(description: str, default_value: int) -> dict[str, Any]:
        """Create an integer property schema with a default value."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.integer_property_with_default")
        return SchemaUtil._property_schema("integer", description, default=default_value)

    @staticmethod
    def optional_integer_property(description: str) -> dict[str, Any]:
        """Create an optional integer property schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.optional_integer_property")
        return SchemaUtil.integer_property(description)

    @staticmethod
    def number_property(description: str) -> dict[str, Any]:
        """Create a number property schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.number_property")
        return SchemaUtil._property_schema("number", description)

    @staticmethod
    def array_property(description: str, items: dict[str, Any] | None = None) -> dict[str, Any]:
        """Create an array property schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.array_property")
        return SchemaUtil._property_schema("array", description, items=items)

    @staticmethod
    def object_property(description: str, properties: dict[str, Any] | None = None) -> dict[str, Any]:
        """Create an object property schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.object_property")
        return SchemaUtil._property_schema("object", description, properties=properties)

    @staticmethod
    def enum_property(description: str, enum_values: list[str]) -> dict[str, Any]:
        """Create an enum property schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.enum_property")
        return SchemaUtil._property_schema("string", description, enum=enum_values)

    @staticmethod
    def create_optional_object_property(description: str, properties: dict[str, Any]) -> dict[str, Any]:
        """Create an optional object property schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.create_optional_object_property")
        return SchemaUtil.object_property(description, properties)

    @staticmethod
    def create_schema(
        properties: dict[str, Any],
        required: list[str] | None = None,
    ) -> dict[str, Any]:
        """Create a complete JSON schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.create_schema")
        schema = {
            "type": "object",
            "properties": properties,
        }
        if required:
            schema["required"] = required
        return schema

    @staticmethod
    def builder() -> SchemaBuilder:
        """Create a schema builder for fluent API."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaUtil.builder")
        return SchemaBuilder()


class SchemaBuilder:
    """Fluent builder for creating JSON schemas."""

    def __init__(self):
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaBuilder.__init__")
        self._properties: dict[str, Any] = {}
        self._required: list[str] = []

    def _add_property(self, name: str, prop: dict[str, Any]) -> SchemaBuilder:
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaBuilder._add_property")
        self._properties[name] = prop
        return self

    def string_property(self, name: str, description: str, default: str | None = None) -> SchemaBuilder:
        """Add a string property."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaBuilder.string_property")
        prop = SchemaUtil.string_property(description)
        if default is not None:
            prop["default"] = default
        return self._add_property(name, prop)

    def boolean_property(self, name: str, description: str, default: bool | None = None) -> SchemaBuilder:
        """Add a boolean property."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaBuilder.boolean_property")
        prop = SchemaUtil.boolean_property(description)
        if default is not None:
            prop["default"] = default
        return self._add_property(name, prop)

    def integer_property(self, name: str, description: str, default: int | None = None) -> SchemaBuilder:
        """Add an integer property."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaBuilder.integer_property")
        prop = SchemaUtil.integer_property(description)
        if default is not None:
            prop["default"] = default
        return self._add_property(name, prop)

    def enum_property(self, name: str, description: str, values: list[str]) -> SchemaBuilder:
        """Add an enum property."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaBuilder.enum_property")
        return self._add_property(name, SchemaUtil.enum_property(description, values))

    def array_property(self, name: str, description: str, items: dict[str, Any] | None = None) -> SchemaBuilder:
        """Add an array property."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaBuilder.array_property")
        return self._add_property(name, SchemaUtil.array_property(description, items))

    def object_property(self, name: str, description: str, properties: dict[str, Any] | None = None) -> SchemaBuilder:
        """Add an object property."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaBuilder.object_property")
        return self._add_property(name, SchemaUtil.object_property(description, properties))

    def required(self, *names: str) -> SchemaBuilder:
        """Mark properties as required."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaBuilder.required")
        self._required.extend(names)
        return self

    def build(self) -> dict[str, Any]:
        """Build the final schema."""
        logger.debug("diag.enter %s", "mcp_utils/schema_util.py:SchemaBuilder.build")
        return SchemaUtil.create_schema(self._properties, self._required or None)
